homeassistant/components/sleepiq/number.py

```diff
@@ -1,15 +1,98 @@
 """Support for SleepIQ SleepNumber firmness number entities."""
-from asyncsleepiq import SleepIQBed, SleepIQSleeper
+from __future__ import annotations
 
-from homeassistant.components.number import NumberEntity
+from collections.abc import Callable, Coroutine
+from dataclasses import dataclass
+from typing import Any, cast
+
+from asyncsleepiq import SleepIQActuator, SleepIQBed, SleepIQSleeper
+
+from homeassistant.components.number import NumberEntity, NumberEntityDescription
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
-from .const import DOMAIN, FIRMNESS
+from .const import ACTUATOR, DOMAIN, ENTITY_TYPES, FIRMNESS, ICON_OCCUPIED
 from .coordinator import SleepIQData
-from .entity import SleepIQSleeperEntity
+from .entity import SleepIQBedEntity
+
+
+@dataclass
+class SleepIQNumberEntityDescriptionMixin:
+    """Mixin to describe a SleepIQ number entity."""
+
+    value_fn: Callable[[Any], float]
+    set_value_fn: Callable[[Any, int], Coroutine[None, None, None]]
+    get_name_fn: Callable[[SleepIQBed, Any], str]
+    get_unique_id_fn: Callable[[SleepIQBed, Any], str]
+
+
+@dataclass
+class SleepIQNumberEntityDescription(
+    NumberEntityDescription, SleepIQNumberEntityDescriptionMixin
+):
+    """Class to describe a SleepIQ number entity."""
+
+
+async def _async_set_firmness(sleeper: SleepIQSleeper, firmness: int) -> None:
+    await sleeper.set_sleepnumber(firmness)
+
+
+async def _async_set_actuator_position(
+    actuator: SleepIQActuator, position: int
+) -> None:
+    await actuator.set_position(position)
+
+
+def _get_actuator_name(bed: SleepIQBed, actuator: SleepIQActuator) -> str:
+    if actuator.side:
+        return f"SleepNumber {bed.name} {actuator.side_full} {actuator.actuator_full} {ENTITY_TYPES[ACTUATOR]}"
+
+    return f"SleepNumber {bed.name} {actuator.actuator_full} {ENTITY_TYPES[ACTUATOR]}"
+
+
+def _get_actuator_unique_id(bed: SleepIQBed, actuator: SleepIQActuator) -> str:
+    if actuator.side:
+        return f"{bed.id}_{actuator.side}_{actuator.actuator}"
+
+    return f"{bed.id}_{actuator.actuator}"
+
+
+def _get_sleeper_name(bed: SleepIQBed, sleeper: SleepIQSleeper) -> str:
+    return f"SleepNumber {bed.name} {sleeper.name} {ENTITY_TYPES[FIRMNESS]}"
+
+
+def _get_sleeper_unique_id(bed: SleepIQBed, sleeper: SleepIQSleeper) -> str:
+    return f"{bed.id}_{sleeper.name}_{FIRMNESS}"
```

Comment:
It looks like we should be using `sleeper.id` here

https://developers.home-assistant.io/docs/entity_registry_index?_highlight=unique&_highlight=id&_highlight=req#unique-id-requirements

Reply:
It looks like it shipped with `sleeper.name` in 2022.3 so we probably need to do a migration.